### PR TITLE
Support multi-valued maintainers

### PR DESCRIPTION
**Pull request by neverpanic**
We have two equal implementations of unobscure_maintainers, but only
a single one should exist. Move the implementation to the macports_util
package where it can be used by port(1) and port1.0, adjust uses in
port(1), drop the implementation from port1.0, because it is unused.

Additionally, add a test that verifies that @github-username works as
expected in base. Note that no changes in the code are required to
support this.

Review request: @macports/portmgr, @mojca, @larryv.

### CODE REVIEW

**Comment by neverpanic** [comment]
I'm trying to avoid adding more code to macports.tcl where I can. That file is already too long.

**Comment by raimue** [comment]
It was more about the namespace than the file it is in, as this moves macports_util from a collection of internal helpers of macports1.0 to the public API.

**Comment by neverpanic** [comment]
Good point, I guess I should move it then.

**Comment by jmroot** [comment]
Splitting some of macports.tcl into different files that each contain some logical subset isn't a bad idea. Even if they're all just sourced by the main file it would probably still improve readability. But that's a bigger project than this PR.

File: src/macports1.0/macports_util.tcl [hunk 1]
```diff
@@ -35,6 +35,36 @@ package provide macports_util 1.0
...
+            if {[string first "@" $maintainer] < 0} {
+                # No "@" in the maintainer, it's either an obfuscated email, or
+                # a keyword
+
+                set colonpos [string first ":" $maintainer]
+                if {$colonpos >= 0} {
+                    # Cut and re-assemble
+                    set maintainer "[string range $maintainer $colonpos+1 end]@[string range $maintainer 0 $colonpos-1]"
+                } elseif {$maintainer ni {"openmaintainer" "nomaintainer"}} {
+                    set maintainer "$maintainer@macports.org"
```
Comment (larryv) [inline src/macports1.0/macports_util.tcl]: Could be
```tcl
append maintainer @macports.org
```

File: src/macports1.0/macports_util.tcl [hunk 2]
```diff
@@ -35,6 +35,36 @@ package provide macports_util 1.0
...
+    ##
+    # Given a list of maintainers as recorded in a Portfile, return an
+    # unobscured version that uses proper email addresses.
+    #
+    # @param list A list of obscured maintainers
+    # @return An unobscured list of maintainers
+    proc unobscure_maintainers {list} {
+        set result {}
+        foreach maintainer $list {
+            if {[string first "@" $maintainer] < 0} {
```
Comment (larryv) [inline src/macports1.0/macports_util.tcl]: I would prefer
```tcl
if {[string first @ $maintainer] >= 0]} {
    continue
}
```
so that the rest of the function could be outdented.

File: src/macports1.0/tests/macports_util.test [hunk 3]
```diff
@@ -30,6 +30,25 @@ test method_wrap {
 } -result "Method wrap successful."
 
 
+test unobscure_maintainers {
+    Unobscure maintainers unit test.
+} -body {
+    if {[macports_util::unobscure_maintainers [list "port"]] ne "port@macports.org"} {
```
Comment (larryv) [inline src/macports1.0/tests/macports_util.test]: Is it necessary to force a pure list here? It's a little noisy 😕

File: src/macports1.0/macports_util.tcl [hunk 4]
```diff
@@ -35,6 +35,36 @@ package provide macports_util 1.0
...
+                # No "@" in the maintainer, it's either an obfuscated email, or
+                # a keyword
+
+                set colonpos [string first ":" $maintainer]
+                if {$colonpos >= 0} {
+                    # Cut and re-assemble
+                    set maintainer "[string range $maintainer $colonpos+1 end]@[string range $maintainer 0 $colonpos-1]"
```
Comment (larryv) [inline src/macports1.0/macports_util.tcl]: Could be
```tcl
lassign [split $maintainer :] domain user
if {$user ne ""} {
    set maintainer ${user}@${domain}
}
```
mainly because I think fussing with indices is noisy.

File: src/macports1.0/macports_util.tcl (hunk 2 above)
Comment (jdberry) [inline src/macports1.0/macports_util.tcl]: But in this case that would result in the maintainer not getting into the result list. You could of course add it to the result before continuing.

**Comment by neverpanic** [comment]
I'll refactor this to include @larryv's comments and the discussion from macports/macports-guide#1.

File: src/macports1.0/macports_util.tcl (hunk 4 above)
Comment (neverpanic) [inline src/macports1.0/macports_util.tcl]: `[split ...]` does not have a maximum number of splits, though, so if you want to get this right it would be

```tcl
set user [join [lassign [split $maintainer :] domain] :]
```

which isn't any shorter either. Nevertheless, that's going to be in the next version of this PR.

**Comment by neverpanic** [comment]
Code now offers a public API in `macports::unobscure_maintainers` with the implementation living in `macports_util`. I've also adjusted `port(1)` to deal with the new format.

As an example:
```tcl
maintainers     {@github-1 handle1} \
                {example.com:email2 @github-2} \
                example.org:email3 \
                @github-4 \
                openmaintainer \
                nomaintainer
```

now produces:
```
Maintainers:          Email: handle1@macports.org, GitHub: github-1
                      Email: email2@example.com, GitHub: github-2
                      Email: email3@example.org
                      GitHub: github-4
                      open maintainer
                      no maintainer
```

File: src/port/port.tcl [hunk 5]
```diff
@@ -2171,9 +2155,40 @@ proc action_info { action portlist opts } {
                 set label "$opt: "
             }
 
+            if {$ropt in {"description" "long_description"}} {
```
Comment (raimue) [inline src/port/port.tcl]: This hunk seems unrelated to unobscuring maintainers.

File: src/port/port.tcl [hunk 6]
```diff
@@ -2171,9 +2155,40 @@ proc action_info { action portlist opts } {
...
+                    if {[info exists maintainer(github)]} {
+                        lappend parts "GitHub: $maintainer(github)"
+                    }
+                    if {[info exists maintainer(keyword)]} {
+                        switch $maintainer(keyword) {
+                            nomaintainer {
+                                lappend parts "no maintainer"
+                            }
+                            openmaintainer {
+                                lappend parts "open maintainer"
```
Comment (raimue) [inline src/port/port.tcl]: I am not sure about these. Maybe I am just used to read them as keywords instead of multiple words… Also seems odd that they have no label.

File: src/port/port.tcl (hunk 5 above)
Comment (neverpanic) [inline src/port/port.tcl]: It is an artifact of the removal of `[join $portinfo($ropt)]` above, which is required because it would otherwise flatten the maintainer list. Not running the `[join ...]` for descriptions breaks ports that use newlines in their descriptions, e.g. `barcode`.

File: src/port/port.tcl (hunk 6 above)
Comment (neverpanic) [inline src/port/port.tcl]: I don't like the current form either, suggestions welcome. `none` and `Policy: openmaintainer` looks much better to me already.

**Comment by raimue** [comment]
We might also need to change the `maintainer:` pseudo-port selector to search in lists.

**Comment by neverpanic** [comment]
Pushed a version with `Maintainer: none` for nomaintainer and `Policy: openmaintainer` for openmaintainer.

I don't agree with changing the `port info` output when `-v` is passed, because that flag is currently meaningless for `port info`, and I don't feel like this is a compelling case to change that.

As for the pseudo-port selector, we have the exact same problem of searching in lists with other fields already, e.g. `depends_*` or `license`. Here are a couple of corner cases to illustrate what would happen:
```
$ rlwrap port-tclsh
% regexp -- "cal" {{@neverpanic cal} {raimue example.com:localpart}}
1
% regexp -- "@neverpanic" {{@neverpanic cal} {raimue example.com:localpart}}
1
% regexp -- "neverpanic" {{@neverpanic cal} {raimue example.com:localpart}}
1
% regexp -- "raimue" {{@neverpanic cal} {raimue example.com:localpart}}
1
% string match "*raimue*" {{@neverpanic cal} {raimue example.com:localpart}}
1
% string compare "{{@neverpanic cal} {raimue example.com:localpart}}" {{@neverpanic cal} {raimue example.com:localpart}}
1
```

**Comment by raimue** [comment]
`port -v info` already changes the dependency information, although this is something most users would not be interested in.

I still do not like the verbosity in the output of `port info`, especially if the only listed entry is an email address, where the "Email:" label looks out of place.

However, as we agreed on the Portfile  syntax, I think we should merge this now and tune the output later.

**Comment by mojca** [comment]
Will we do a mass commit for this change? I would expect a mass commit that adds github usernames to all developers where we know that info.

**Comment by raimue** [comment]
I would like to see it changed at once, but mass commits are a problem for the buildbot, as we discussed when we wanted to remove the `$Id$` tags.

**Comment by neverpanic** [comment]
Remember that this requires a new base release plus a week that we allow for users to install the new base version, so we have time to figure out how to do this, e.g. by implementing the successcache.

**Comment by mojca** [comment]
We need to fix the parser on the buildbot setup.
